File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-cpp/tests/dispatch_queue_test.cpp
// Copyright (C) 2026 Antony Stubbs and contributors
//
// THE IN-FLIGHT CEILING, which is the rule three of the first five clients in this fan-out
// implemented wrongly from the same text. Every test here is written against the authoring guide's
// §3 rule 2 - "max_concurrency bounds the records this client has been dispatched and has not yet
// reported, QUEUED PLUS EXECUTING" - and the third one is the guide's own worked example, which is
// the only shape that discriminates a correct client from the defect.
//
// PROVEN RED BEFORE GREEN: with DispatchQueue::admit changed to compare `queued_.size()` against the
// ceiling instead of `unresolved_` - the exact defect the guide describes - "a record out with an
// executor still occupies the ceiling" and "only a verdict frees a slot" both fail, while
// "overflowing the ceiling ..." (one wave larger than the ceiling) still PASSES. That is the whole
// argument for why the wave-sized test is not enough on its own.

#include "dispatch_queue.h"
#include "error.h"
#include "test_support.h"

namespace {

namespace pcp = parallelconsumer::proxy;
namespace v1 = parallelconsumer::proxy::v1;

v1::DispatchRecord record_at(std::int64_t offset) {
    v1::DispatchRecord dispatched;
    dispatched.mutable_token()->set_record_id("record-" + std::to_string(offset));
    dispatched.mutable_token()->set_epoch(1);
    dispatched.mutable_record()->set_offset(offset);
    dispatched.set_attempt(1);
    return dispatched;
}

PCP_TEST(fifo_within_a_wave, "a wave is handed out in record order") {
    pcp::DispatchQueue queue;
    queue.configure(3);
    for (std::int64_t offset = 0; offset < 3; ++offset) {
        queue.admit(record_at(offset));
    }

    std::vector<std::int64_t> handed;
    v1::DispatchRecord taken;
    queue.close();  // so take() returns false rather than blocking once drained
    while (queue.take(taken)) {
        handed.push_back(taken.record().offset());
    }

    PCP_CHECK_EQ(handed.size(), std::size_t{3});
    PCP_CHECK_EQ(handed[0], std::int64_t{0});
    PCP_CHECK_EQ(handed[1], std::int64_t{1});
    PCP_CHECK_EQ(handed[2], std::int64_t{2});
}

PCP_TEST(a_wave_larger_than_the_ceiling_overflows,
         "a wave larger than the ceiling is a protocol violation naming the counts and the token") {
    pcp::DispatchQueue queue;
    queue.configure(2);
    queue.admit(record_at(0));
    queue.admit(record_at(1));

    bool refused = false;
    try {
        queue.admit(record_at(2));
    } catch (const pcp::ProtocolError& violation) {
        refused = true;
        const std::string message = violation.what();
        PCP_CHECK_CONTAINS(message, "2 were already unresolved");
        PCP_CHECK_CONTAINS(message, "max_concurrency of 2");
        PCP_CHECK_CONTAINS(message, "cancelled");
        // The token, rendered as it arrived. Opacity forbids DERIVING from it, not printing it, and
        // the specification's overflow contract requires it - no other client in this fan-out does
        // this yet.
        PCP_CHECK_CONTAINS(message, "record-2");
        PCP_CHECK_CONTAINS(message, "epoch: 1");
    }
    PCP_CHECK(refused);
}

/// THE GUIDE'S OWN WORKED EXAMPLE, and the only shape the wave-sized test above cannot reach:
/// ceiling 3, two executors holding A and B, C still queued, and a fourth record arriving. Counting
/// queued records alone, D fits - the deque holds one and has two slots free - and the overflow the
/// guide describes at length is undetectable.
PCP_TEST(executing_records_still_occupy_the_ceiling,
         "a record out with an executor still occupies the ceiling") {
    pcp::DispatchQueue queue;
    queue.configure(3);
    queue.admit(record_at(0));
    queue.admit(record_at(1));
    queue.admit(record_at(2));

    v1::DispatchRecord taken;
    PCP_CHECK(queue.take(taken));  // executor-1 takes A
    PCP_CHECK(queue.take(taken));  // executor-2 takes B
    PCP_CHECK_EQ(queue.depth(), std::size_t{1});
    PCP_CHECK_EQ(queue.unresolved(), 3);

    bool refused = false;
    try {
        queue.admit(record_at(3));
    } catch (const pcp::ProtocolError& violation) {
        refused = true;
        PCP_CHECK_CONTAINS(std::string(violation.what()), "3 were already unresolved");
    }
    PCP_CHECK(refused);
}

PCP_TEST(only_a_verdict_frees_a_slot, "reporting a record is what makes room for the next") {
    pcp::DispatchQueue queue;
    queue.configure(1);
    queue.admit(record_at(0));
    v1::DispatchRecord taken;
    PCP_CHECK(queue.take(taken));  // the only executor takes it
    PCP_CHECK_EQ(queue.depth(), std::size_t{0});

    bool refused = false;
    try {
        queue.admit(record_at(1));
    } catch (const pcp::ProtocolError&) {
        refused = true;
    }
    PCP_CHECK(refused);

    queue.settle();  // the report
    bool admitted = true;
    try {
        queue.admit(record_at(1));
    } catch (const pcp::ProtocolError&) {
        admitted = false;
    }
    PCP_CHECK(admitted);
}

PCP_TEST(discarding_is_a_verdict_too, "records discarded at shutdown stop counting against the ceiling") {
    pcp::DispatchQueue queue;
    queue.configure(2);
    queue.admit(record_at(0));
    queue.admit(record_at(1));

    PCP_CHECK_EQ(queue.discard_queued(), 2);
    PCP_CHECK_EQ(queue.unresolved(), 0);
}

PCP_TEST(stop_handout_leaves_queued_records_alone, "stopping hand-out stops the queue, not the ceiling") {
    pcp::DispatchQueue queue;
    queue.configure(2);
    queue.admit(record_at(0));
    queue.stop_handout();

    v1::DispatchRecord taken;
    PCP_CHECK(!queue.take(taken));
    // The record is still unresolved: hand-out stopping is not a verdict, and inventing one for work
    // this client did not do is exactly what the protocol forbids.
    PCP_CHECK_EQ(queue.unresolved(), 1);
}

}  // namespace

File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-go/parallelconsumer/dispatch_queue_test.go
// Copyright (C) 2026 Antony Stubbs and contributors

package parallelconsumer

import (
	"context"
	"fmt"
	"strings"
	"testing"

	proxyv1 "github.com/astubbs/parallel-consumer/parallel-consumer-proxy-clients/parallel-consumer-proxy-client-go/gen/parallelconsumer/proxy/v1"
)

// dispatchOf builds a wave of n records with consecutive record ids, in order. The record ids are
// the only thing the tests read back, because FIFO is the one property expressible as an order.
func dispatchOf(n int, from int) *proxyv1.Dispatch {
	d := &proxyv1.Dispatch{}
	for i := 0; i < n; i++ {
		id := fmt.Sprintf("record-%d", from+i)
		topic, offset := "a-topic", int64(from+i)
		d.Records = append(d.Records, &proxyv1.DispatchRecord{
			Token:  &proxyv1.Token{RecordId: &id},
			Record: &proxyv1.Record{Topic: &topic, Offset: &offset, Value: []byte("v")},
		})
	}
	return d
}

// takeFromQueue is an executor taking a record, and nothing more: it leaves the queue and starts
// running, which is precisely the moment a queue-length bound wrongly frees a slot.
func takeFromQueue(t *testing.T, c *Client) *proxyv1.DispatchRecord {
	t.Helper()
	select {
	case rec := <-c.queue:
		return rec
	default:
		t.Fatal("an executor found the queue empty")
		return nil
	}
}

// TestARecordOutWithAnExecutorStillOccupiesTheCeiling is the client-authoring guide's own worked
// example, and the shape that discriminates: max_concurrency 3, A B and C admitted, two of them
// taken by executors so the queue is nearly empty while the ceiling is full, nothing reported, and
// a fourth record arriving. A client bounding its QUEUE has two free slots and admits D without a
// murmur - which is the defect this test exists for, and why the wave-larger-than-the-ceiling test
// below cannot stand in for it.
func TestARecordOutWithAnExecutorStillOccupiesTheCeiling(t *testing.T) {
	c := newTestClient(newFakeStream(), 3, 2)

	if err := c.enqueue(dispatchOf(3, 1)); err != nil {
		t.Fatalf("a wave exactly filling the ceiling was refused: %v", err)
	}
	takeFromQueue(t, c) // executor-1 takes A
	takeFromQueue(t, c) // executor-2 takes B

	if got := c.unresolved.Load(); got != 3 {
		t.Fatalf("unresolved was %d after two records left the queue for executors, want 3 - "+
			"leaving the queue is not a verdict and does not free a slot", got)
	}

	err := c.enqueue(dispatchOf(1, 4))
	if err == nil {
		t.Fatal("a fourth record was admitted while three were unresolved - the proxy exceeded the " +
			"ceiling it declared itself and this client could not tell")
	}
	for _, want := range []string{"3 were already unresolved", "max_concurrency of 3", "record_id"} {
		if !strings.Contains(err.Error(), want) {
			t.Errorf("the violation did not name %q: %v", want, err)
		}
	}
}

// TestReportingIsWhatFreesTheSlot is the other half of the same rule: only a verdict makes room.
func TestReportingIsWhatFreesTheSlot(t *testing.T) {
	c := newTestClient(newFakeStream(), 1, 1)

	if err := c.enqueue(dispatchOf(1, 1)); err != nil {
		t.Fatalf("the first record was refused: %v", err)
	}
	takeFromQueue(t, c)

	if err := c.enqueue(dispatchOf(1, 2)); err == nil {
		t.Fatal("a second record was admitted while the first was still executing")
	}

	c.settle()
	if err := c.enqueue(dispatchOf(1, 3)); err != nil {
		t.Fatalf("the slot was not freed by the verdict: %v", err)
	}
}

// TestAReportedRecordSettlesOnTheWayOut runs the real report path, because the accounting is only
// as good as the place it is called from: the decrement sits in a defer so that an executor dying
// between the send and the decrement cannot skip it, and a defer is invisible to a test that only
// calls settle by hand.
func TestAReportedRecordSettlesOnTheWayOut(t *testing.T) {
	stream := newFakeStream()
	c := newTestClient(stream, 2, 1)

	if err := c.enqueue(dispatchOf(1, 1)); err != nil {
		t.Fatalf("the record was refused: %v", err)
	}
	rec := takeFromQueue(t, c)

	c.runOne(context.Background(), func(context.Context, InboundRecord) (Outcome, error) {
		if got := c.unresolved.Load(); got != 1 {
			t.Errorf("unresolved was %d while the record was executing, want 1", got)
		}
		return Succeed(), nil
	}, rec)

	if got := c.unresolved.Load(); got != 0 {
		t.Errorf("unresolved was %d after the report was sent, want 0", got)
	}
	if sent := stream.sentMessages(); len(sent) != 1 || sent[0].GetReport() == nil {
		t.Fatalf("want exactly one Report on the stream, got %d message(s): %v", len(sent), sent)
	}
}

// TestDiscardingTheQueueAtShutdownFreesItsSlots covers the second of the guide's decrement points
// this client can reach: a session without the `shutdown` capability drops its queued records and
// reports nothing for them, so nothing else will ever free them.
func TestDiscardingTheQueueAtShutdownFreesItsSlots(t *testing.T) {
	c := newTestClient(newFakeStream(), 3, 1)

	if err := c.enqueue(dispatchOf(3, 1)); err != nil {
		t.Fatalf("the wave was refused: %v", err)
	}
	takeFromQueue(t, c) // one is out with an executor; two are still queued

	c.discardQueue()

	if got := c.unresolved.Load(); got != 1 {
		t.Errorf("unresolved was %d after discarding two queued records, want 1 - the executing "+
			"record still holds its slot until it reports", got)
	}
}

// TestSettlingBelowZeroIsSaturating keeps a double settle from wrapping the count into a ceiling
// nothing could ever overflow, which would turn this check back into one that cannot fire.
func TestSettlingBelowZeroIsSaturating(t *testing.T) {
	c := newTestClient(newFakeStream(), 1, 1)

	c.settle()
	c.settle()

	if got := c.unresolved.Load(); got != 0 {
		t.Fatalf("unresolved was %d after settling an empty count, want 0", got)
	}
	if err := c.enqueue(dispatchOf(1, 1)); err != nil {
		t.Fatalf("the queue would not admit a record after a double settle: %v", err)
	}
	if err := c.enqueue(dispatchOf(1, 2)); err == nil {
		t.Fatal("the ceiling stopped firing after a double settle")
	}
}

// TestAWaveLargerThanTheCeilingOverflows is the test shape the guide names as the one that LOOKS
// like the negative control and is not: a single wave of four against a ceiling of three trips any
// bound at all, including a bound on the queue's own length, so it passes identically against the
// defect it appears to be named for. Kept because the case is real, labelled because it proves
// nothing on its own.
func TestAWaveLargerThanTheCeilingOverflows(t *testing.T) {
	c := newTestClient(newFakeStream(), 3, 2)

	if err := c.enqueue(dispatchOf(4, 1)); err == nil {
		t.Fatal("a wave of four against a ceiling of three was admitted")
	}
}

// TestHandOutIsFifo is rule 3: by arrival, and within one wave by the wave's own order.
func TestHandOutIsFifo(t *testing.T) {
	c := newTestClient(newFakeStream(), 4, 1)

	if err := c.enqueue(dispatchOf(2, 1)); err != nil {
		t.Fatalf("the first wave was refused: %v", err)
	}
	if err := c.enqueue(dispatchOf(2, 3)); err != nil {
		t.Fatalf("the second wave was refused: %v", err)
	}
	for want := 1; want <= 4; want++ {
		if got := takeFromQueue(t, c).GetToken().GetRecordId(); got != fmt.Sprintf("record-%d", want) {
			t.Fatalf("hand-out order reached %q where FIFO wants record-%d", got, want)
		}
	}
}

// sentMessages is a copy, so an assertion never reads the slice the client is appending to.
func (f *fakeStream) sentMessages() []*proxyv1.ClientMessage {
	f.mu.Lock()
	defer f.mu.Unlock()
	return append([]*proxyv1.ClientMessage(nil), f.sent...)
}
